### PR TITLE
Add typing to state machine signatures

### PR DESCRIPTION
**Pull request by collijk**

### CODE REVIEW

**Review by mattkappel — APPROVED** [review]

File: src/vivarium/framework/state_machine.py [hunk 1]
```diff
@@ -108,45 +110,46 @@ class Transition:
...
 
     """
     def __init__(self,
-                 input_state,
-                 output_state,
-                 probability_func=lambda index: pd.Series(1, index=index),
+                 input_state: 'State',
+                 output_state: 'State',
+                 probability_func: Callable[[pd.Index], pd.Series] = lambda index: pd.Series(1, index=index),
                  triggered=Trigger.NOT_TRIGGERED):
```
Comment (rmudambi) [inline src/vivarium/framework/state_machine.py:126]: Shouldn't `triggered` have a type declaration?

File: src/vivarium/framework/state_machine.py [hunk 2]
```diff
@@ -108,45 +110,46 @@ class Transition:
 
     Parameters
     ----------
-    input_state: State
```
Comment (rmudambi) [inline src/vivarium/framework/state_machine.py:111]: In line 97, shouldn't `_process_trigger()` have both input and output types specified?

File: src/vivarium/framework/state_machine.py [hunk 3]
```diff
@@ -158,38 +161,40 @@ def probability(self, index):
         return activated.append(null)
 
     def __repr__(self):
```
Comment (rmudambi) [inline src/vivarium/framework/state_machine.py:163]: Output type needed?

File: src/vivarium/framework/state_machine.py [hunk 4]
```diff
@@ -220,32 +225,41 @@ def transition_effect(self, index: pd.Index, event_time: pd.Timestamp, populatio
...
-    def cleanup_effect(self, index, event_time):
+    def cleanup_effect(self, index: pd.Index, event_time: 'Time') -> None:
         self._cleanup_effect(index, event_time)
 
-    def add_transition(self, output,
-                       probability_func=lambda index: np.ones(len(index), dtype=float),
-                       triggered=Trigger.NOT_TRIGGERED):
+    def add_transition(self, output: 'State',
+                       probability_func: Callable[[pd.Index], pd.Series] = lambda index: pd.Series(1.0, index=index),
+                       triggered=Trigger.NOT_TRIGGERED) -> Transition:
```
Comment (rmudambi) [inline src/vivarium/framework/state_machine.py:233]: Specify type of `triggered`

File: src/vivarium/framework/state_machine.py [hunk 5]
```diff
@@ -220,32 +225,41 @@ def transition_effect(self, index: pd.Index, event_time: pd.Timestamp, populatio
...
-    def _transition_side_effect(self, index, event_time):
+    def _transition_side_effect(self, index: pd.Index, event_time: 'Time') -> None:
         pass
 
-    def _cleanup_effect(self, index, event_time):
+    def _cleanup_effect(self, index: pd.Index, event_time: 'Time') -> None:
         pass
 
     def __repr__(self):
-        return f'State({self.state_id})'
```
Comment (rmudambi) [inline src/vivarium/framework/state_machine.py:248]: Specify output type? Same in line 271

**Review by rmudambi — APPROVED** [review]

**Review by mattkappel — APPROVED** [review]

**Review by KjellSwedin — APPROVED** [review]

**Pull request merged**